ui/src/app/utils/cluster-initializer.tsx

```diff
@@ -0,0 +1,136 @@
+      this.currentCluster.clusterNamespace, this.currentCluster.clusterName, {signal: this.abortSignal});
+  }
+
+  private async deleteCluster(): Promise<void> {
+    if (this.deleteCount >= MAX_DELETE_COUNT) {
+      throw new Error('Reached max cluster delete count');
+    }
+    await clusterApi().deleteCluster(this.workspaceNamespace, {signal: this.abortSignal});
+    this.deleteCount++;
+  }
+
+  private isClusterRunning(): boolean {
+    return this.currentCluster && this.currentCluster.status === ClusterStatus.Running;
+  }
+
+  private isClusterStopped(): boolean {
+    return this.currentCluster && this.currentCluster.status === ClusterStatus.Stopped;
+  }
+
+  private isClusterErrored(): boolean {
+    return this.currentCluster && this.currentCluster.status === ClusterStatus.Error;
+  }
+
+  async initialize(): Promise<Cluster> {
+    console.log('Initializing cluster', this.workspaceNamespace);
+    if (this.isInitializing) {
+      return Promise.reject('Initialization is already in progress');
+    }
+    this.initializeStartTime = new Date().getTime();
+
+    // Overall strategy: continue polling the get-cluster endpoint, with capped exponential backoff,
+    // until we either reach our goal state (a RUNNING cluster) or run up against the overall
+    // timeout threshold.
+    while (!this.isClusterRunning()) {
+      if (this.abortSignal.aborted) {
+        console.log('abortSignal received!');
+        return Promise.reject('Request was aborted.');
+      }
+      if (new Date().getTime() - this.initializeStartTime > INITIALIZE_TIMEOUT_MS) {
+        console.log('Took longer than max init timeout');
+        return Promise.reject('Initialization attempt took longer than the max time allowed.');
+      }
+
+      // Fetch the cluster and update status.
+      try {
+        console.log('Getting cluster');
+        this.currentCluster = await this.getCluster();
+        console.log('Cluster status', this.currentCluster.status);
+        this.onStatusUpdate(this.currentCluster.status);
+      } catch (e) {
+        console.log('getCluster error', e);
+        if (isAbortError(e)) {
+          return Promise.reject();
+        }
+        // If we received a NOT_FOUND error, we need to create a cluster for this workspace.
+        this.currentCluster = await this.createCluster();
+      }
+
+      // Resolve the promise if this cluster is all set.
+      if (this.isClusterRunning()) {
+        console.log('Cluster is running, resolving promise');
+        return this.currentCluster;
+      }
+
+      if (this.isClusterStopped()) {
+        console.log('Cluster is stopped, resuming');
+        await this.resumeCluster();
+      }
+
+      // If cluster is in error state, delete it so it can be re-created at the next poll loop.
+      if (this.isClusterErrored()) {
+        // TODO: report errors to Stackdriver here.
+        console.log('Cluster error -- deleting for retry');
+        await this.deleteCluster();
```

Comment:
Per ticket discussion, I might move this above the loop so it only happens on initialization - errors encountered later could always throw and require user intervention. I think this limits the risk of concurrent pollers causing cluster churn - delete is the only really problematic/destructive action here. The rest should be relatively safe for concurrent requests.

Reply:
Yes makes sense, I surmised as much from the changes - thanks.